tsfc/driver.py

```diff
@@ -377,7 +377,11 @@ def compile_expression_dual_evaluation(expression, to_element, *,
             quad_rule = QuadratureRule(point_set, to_element._weights)
             config["quadrature_rule"] = quad_rule
 
+        # evaluate expression at the points specified by point_set
         expr, = fem.compile_ufl(expression, **config, point_sum=False)
+        # the point set free indices should now be free indices of the compiled
+        # expression
+        assert set(point_set.indices) <= set(expr.free_indices)
```

Comment:
@wence- there was some discussion yesterday of adding some kind of assertion along these lines, but I can't work out what it should be. Given that the point set indices _can_ disappear, and we don't know what other free indices there might be at this point, what's there left to assert?

Reply:
What other free indices might there be?

Reply:
I think `expr.free_indices \in point_set.indices`?

Reply:
I _think_ if your expression has an argument there is a free index for that. Might be remembering wrong though

Reply:
Oh yeah, so what we're expecting is `expr.free_indices <= set(chain(point_set.indices, *argument_multiindices)))` I think

Reply:
See https://github.com/firedrakeproject/tsfc/pull/241